java/vector/src/main/java/org/apache/arrow/vector/sort/VariableWidthVectorSorter.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.sort;
+
+import static org.apache.arrow.vector.BaseVariableWidthVector.OFFSET_WIDTH;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.arrow.vector.BaseVariableWidthVector;
+import org.apache.arrow.vector.BitVectorHelper;
+
+import io.netty.buffer.ArrowBuf;
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * Default sorter for variable-width vectors.
+ * It is an out-of-place sort, with time complexity O(n*log(n)).
+ * @param <V> vector type.
+ */
+public class VariableWidthVectorSorter<V extends BaseVariableWidthVector> implements VectorSorter<V> {
+
+  @Override
+  public V sort(V srcVector, VectorValueComparator<V> comparator) {
+    comparator.attachVector(srcVector);
+
+    // create the output vector
+    V dstVector = comparator.newVector(srcVector.getAllocator());
+    dstVector.allocateNew(srcVector.getByteCapacity(), srcVector.getValueCount());
+
+    // buffers referenced in the sort
+    ArrowBuf srcValueBuffer = srcVector.getDataBuffer();
+    ArrowBuf srcOffsetBuffer = srcVector.getOffsetBuffer();
+    ArrowBuf dstValidityBuffer = dstVector.getValidityBuffer();
+    ArrowBuf dstValueBuffer = dstVector.getDataBuffer();
+    ArrowBuf dstOffsetBuffer = dstVector.getOffsetBuffer();
+
+    // sort value indices
+    List<Integer> sortedIndices = IntStream.range(0, srcVector.getValueCount()).boxed().collect(Collectors.toList());
```

Comment:
I think there might be some value of separating out the functionality into two components (and then potentially wrapping the two):
1.  Returning a ValueVector of indices indicating the sort order.
2.  Given a ValueVector of indices a utlity to copy the data from those indices into a new ValueVector (and a utility to swap them in place for fixed width types.

What do you think?

Reply:
This is a good point. 
In the new commits, I have extracted the logic for sorting indices into a separate utility class named IndexSorter, which is based on quick sort.

Also, please note that not all sorting algorithms are based on the above 2 step paradigm. Some algorithms can be implemented by the 2 step paradigm, but the performance is not optimal.